benchmark/benchmark-atomic.cpp

```diff
@@ -0,0 +1,320 @@
+    using policy = RAJA::omp_parallel_for_exec;;
+    static std::string PolicyName() {
+        std::stringstream ss;
+        ss << "OpenMP execution";
+        return ss.str();
+    }
+};
+
+/// Functor wrapping the desul implementation.  Wrapping the desul call ensure an identical signature with
+/// RAJA's implementations.  Wrapping the call in an functor allows simple type deduction for printing
+/// from within the benchmark.
+template<typename T, typename ReturnType, typename DesulAtomicSignature<ReturnType, T*, const T, T>::type atomic_impl>
+struct atomicWrapperDesulTernary {
+    /// Call operator overload template that allows invoking DESUL atomic with a (int*)(T*, T) signature
+    RAJA_HOST_DEVICE ReturnType operator()(T * acc, T value) const {
+        return atomic_impl(acc, value, T(1), raja_default_desul_order{},
+                        raja_default_desul_scope{});
+    }
+};
+
+template<typename T, typename ReturnType, typename DesulAtomicSignature<ReturnType, T*, const T>::type atomic_impl>
+struct atomicWrapperDesulBinary {
+    /// Call operator overload template that allows invoking DESUL atomic with a (int*)(T*, T) signature
+    RAJA_HOST_DEVICE ReturnType operator()(T * acc, T value) const {
+        return atomic_impl(acc, value, raja_default_desul_order{},
+                        raja_default_desul_scope{});
+    }
+};
+
+/// Unary wrapper variant for increment and decrement benchmarks.
+template<typename T, typename ReturnType, typename DesulAtomicSignature<ReturnType, T*>::type atomic_impl>
+struct atomicWrapperDesulUnary {
+    RAJA_HOST_DEVICE ReturnType operator()(T* acc, T) const {
+        return atomic_impl(acc, raja_default_desul_order{},
+                        raja_default_desul_scope{});
+    }
+};
+
+template<typename T>
+class IsDesul : public std::false_type {};
+
+template<typename T, typename ReturnType, typename DesulAtomicSignature<ReturnType, T, T*>::type atomic_impl>
+class IsDesul<atomicWrapperDesulBinary<T, ReturnType, atomic_impl>> : public std::true_type {};
+
+template<typename T, typename ReturnType, typename DesulAtomicSignature<ReturnType, T>::type atomic_impl>
+class IsDesul<atomicWrapperDesulUnary<T, ReturnType, atomic_impl>> : public std::true_type {};
+
+template<typename AtomicImplType>
+std::string GetImplName (const AtomicImplType& impl) {
+    if (IsDesul<AtomicImplType>::value) {
+        return "Desul atomic";
+    } else {
+        return "RAJA atomic";
+    }
+}
+
+template <class ExecPolicy, typename AtomicType, bool test_array, typename AtomicImplType>
+void TimeAtomicOp(const AtomicImplType& atomic_impl, uint64_t N, uint64_t num_iterations = 4, int array_size = 100, bool print_to_output = true) {
+    RAJA::Timer timer;
+
+    // Allocate memory
+    AtomicType* device_value = nullptr;
+    int len_array = test_array ? array_size : 1;
+    camp::resources::Resource resource {RAJA::resources::get_resource<typename ExecPolicy::policy>::type::get_default()};
```

Comment:
Is there a reason to not just use the typed resource directly?
```suggestion
    auto resource {RAJA::resources::get_resource<typename ExecPolicy::policy>::type::get_default()};
```